lib/vector-core/src/stream/batcher.rs

```diff
@@ -1,742 +1,261 @@
-    use crate::stream::batcher::{Batcher, ExpirationQueue};
-    use crate::time::KeyedTimer;
-    use futures::{stream, Stream};
-    use pin_project::pin_project;
-    use proptest::prelude::*;
-    use std::collections::{HashMap, HashSet};
-    use std::num::{NonZeroU8, NonZeroUsize};
-    use std::pin::Pin;
-    use std::task::{Context, Poll};
-    use std::time::Duration;
-    use tokio::pin;
-    use tokio::time::advance;
+    /// returns true iff it is not possible for another item to fit in the batch
+    fn is_batch_full(&self) -> bool {
+        self.batch.len() >= self.batch_item_limit || self.current_size >= self.batch_size_limit
+    }
 
-    #[derive(Debug)]
-    /// A test keyed timer
-    ///
-    /// This timer implements `KeyedTimer` and is rigged up in such a way that
-    /// it doesn't _actually_ tell time but instead uses a set of canned
-    /// responses for whether deadlines have elapsed or not. This allows us to
-    /// include the notion of time in our property tests below.
-    struct TestTimer {
-        responses: Vec<Poll<Option<u8>>>,
-        valid_keys: HashSet<u8>,
+    fn take_batch(self: Pin<&mut Self>) -> Vec<S::Item> {
+        let this = self.project();
+        *this.current_size = 0;
+        std::mem::take(this.batch)
     }
 
-    impl TestTimer {
-        fn new(responses: Vec<Poll<Option<u8>>>) -> Self {
-            Self {
-                responses,
-                valid_keys: HashSet::new(),
-            }
-        }
+    fn push_item(self: Pin<&mut Self>, item: S::Item, item_size: usize) {
+        let this = self.project();
+        this.batch.push(item);
+        *this.current_size += item_size;
     }
 
-    impl KeyedTimer<u8> for TestTimer {
-        fn clear(&mut self) {
-            self.valid_keys.clear();
+    fn reset_timer(self: Pin<&mut Self>) {
+        let timeout = self.timeout;
+        let mut this = self.project();
+        if let MaybeProj::Some(timer) = this.timer.as_mut().project() {
+            timer.reset(tokio::time::Instant::now() + timeout);
+        } else {
+            this.timer.set(Maybe::Some(tokio::time::sleep(timeout)));
         }
+    }
+}
 
-        fn insert(&mut self, item_key: u8) {
-            self.valid_keys.insert(item_key);
-        }
+impl<S, I> Stream for Batcher<S, I>
+where
+    S: Stream,
+    I: ItemBatchSize<S::Item>,
+{
+    type Item = Vec<S::Item>;
 
-        fn poll_expired(&mut self, _cx: &mut Context) -> Poll<Option<u8>> {
-            match self.responses.pop() {
-                Some(Poll::Pending) => unreachable!(),
-                None | Some(Poll::Ready(None)) => Poll::Ready(None),
-                Some(Poll::Ready(Some(k))) => {
-                    if self.valid_keys.contains(&k) {
-                        Poll::Ready(Some(k))
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            match self.as_mut().project().stream.poll_next(cx) {
```

Comment:
At a high level, just a cleanliness, non-blocking comment: might want to just project self at the very top here to avoid all of the `.as_mut()` calls below?

It definitely feels a little weird to see inherent methods on `Batcher` take a pinned method receiver, vs constraining that portion to only where we need it.

Reply:
I agree it's a bit ugly as is, but unless I'm missing something, I don't see a reasonable way to clean it up.

If I `.project()` at the top, that moves out `self` which prevent me from calling any functions on it. I think the functions keep the code cleaner than removing the `.as_mut().project()`.